traceutil/roundtrip.go

```diff
@@ -0,0 +1,191 @@
+package traceutil
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"net/http/httptrace"
+	"sync"
+	"time"
+)
+
+// RoundTripResult is a set of timestamps set at various stages of an outgoing
+// HTTP request intended to be passed to a FinishFunc passed in to RoundTrip.
+type RoundTripResult struct {
+	Started              time.Time
+	WroteHeaders         time.Time
+	WroteRequest         time.Time
+	GotFirstResponseByte time.Time
+	ReadHeaders          time.Time
+	ReadResponse         time.Time
+
+	BytesWritten int // TODO: implemenent
+	BytesRead    int
+
+	// The request traced. It is not safe to modify nor read from its body
+	// or headers.
+	Request *http.Request
+
+	// The response traced. It is not safe to modify nor read from its body
+	// of headers.  It is nil if no response was recieved.
+	Response *http.Response
+
+	Err error
+}
+
+func propagateHTTP(req *http.Request) {
+	// TODO: implement
+}
+
+type traceRoundTripper struct {
+	inner    http.RoundTripper
+	finished FinishFunc
+}
+
+// RoundTrip returns a http.RoundTripper that wraps inner for purposes of
+// recording timestamps to a RoundTripResult which is passed to finished on
+// completion or error of the returned http.RoundTripper.
+func RoundTrip(inner http.RoundTripper, finished FinishFunc) http.RoundTripper {
+	return traceRoundTripper{inner: inner, finished: finished}
+}
+
+type FinishFunc func(result *RoundTripResult)
+
+// FinishBasic records a single span for the request representing the from
+// first request byte written to last response byte read.
+func FinishBasic() FinishFunc {
+	panic("TODO")
+}
+
+// FinishDetail records sevaral detailed spans for the request representing the
+// from first request byte written to last response byte read.
+//
+// The spans are:
+//
+//    |==RoundTrip=======================================================|
+//    |==WroteHeader==                                                   |
+//    |               ==WroteBody==                                      |
+//    |                                  ==ReadFirstByte==               |
+//    |                                                   ==ReadHeaders==|
+//    |==================================================================|
+//
+func FinishDetail() FinishFunc {
+	panic("TODO")
+}
+
+func (r traceRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	var (
+		trace   = newClientTrace()
+		ctx     = httptrace.WithClientTrace(req.Context(), trace.client)
+		timeNow = timeNowFromContext(ctx)
+	)
+	req = req.WithContext(ctx)
+
+	// TODO(bmizerany): check if already configured for propagation as to
+	// not override anything custom? We could do this with a key in the
+	// req.Context.
+	propagateHTTP(req)
+
+	trace.Request = req
+	trace.Started = timeNow()
+
+	res, err := r.inner.RoundTrip(req)
+	if err != nil {
```

Comment:
I had a note for that and forgot to submit the review. I'll reaudit the code soon. I just wanted to get the basics out there to discuss API.